### ISSUE TITLE
gdrive metadata should override mongo-stub metadata on userpage

### ISSUE THREAD

**Issue by ericscheid** [naturalcrit/homebrewery#2692]
### Renderer

v3

### Browser

Chrome

### Operating System

MacOS

### What happened?

On the userpage, the metadata displayed should be from either the mongo record (stub or full), or the properties retrieved from the GDrive metadata API, with precedence given to the gdrive properties.

The precedence should be something like `metadata codefence in the text` > `Gdrive file props` > `mongo stub props`, with some brew props only coming from the `mongo` record (e.g. brew `views` count).

----

Currently, it appears that on the userpage the title and last-modified date (at least) is being sourced from the mongo stub record and not from the gdrive metadata. That is, if the gdrive file is modified, and renamed, then the userpage does not reflect that. 
Renaming the gdrive file does not change the title displayed on the userpage. 

<img width="936" alt="image" src="https://user-images.githubusercontent.com/2605467/220851660-6d87e942-a2dd-44ff-b72f-249abab38a6b.png">

<img width="348" alt="image" src="https://user-images.githubusercontent.com/2605467/220850609-5aa475d8-8a33-402d-a6be-e6e6207cfb6e.png">

Loading the brew in question into the homebrewery editor will reveal the updated title and description, sourced from the codefenced metadata section of the brew contents.

<img width="863" alt="image" src="https://user-images.githubusercontent.com/2605467/220850711-bb7870d9-247d-444f-b8a2-9295ff69c6ad.png">

Saving the brew will rename the gdrive file to match the title in the metadata panel (which being loaded from the codefenced metadata content could be different from the new filename).

Similarly, the share link of the brew uses the title from the brew contents metadata.

<img width="893" alt="image" src="https://user-images.githubusercontent.com/2605467/220851168-448c239f-29ff-4942-be4d-53d22b91699d.png">

## Proposals

1. the userpage should display title and last-modified from the gdrive metadata, and whatever other valid metadata from gdrive file metadata (i.e. without needing to fetch the full file content)
2. [optional] indicate if the gdrive last-modified is different from the mongo last-modified
3. [optional-plus] one-click "refresh metadata" button if difference is detected (additionally this could even indicate gdrive brews that don't even have a mongo stub, with the one-click button thus creating the stub)

**Comment by dbolack-ab** [comment]
Stubs should sync with the brew data so that they can be properly loaded without loading the brew.

**Comment by calculuschild** [comment]
> Stubs should sync with the brew data so that they can be properly loaded without loading the brew.

Once the file is loaded in HB, they do sync up.

The issue here is when the Google Drive file is edited externally from The Homebrewery. In that case, the Google Drive file should take precedence if it doesn't match the stub in Mongo.

**Comment by dbolack-ab** [comment]
The problem with that, with the example above, is it would require opening every brew to display the stub metadata correctly in order to build the cards correctly. This seems wasteful.

If this is absolutely desirable, I think we could get away with only loading the Gdrive metadata if the GDrive datestamp is later than the stub lastUpdated.

_I'm of the mind that so long as the stub is updated when the brew is loaded or saved, it's fine that the data is wrong because someone edited the data from outside._

**Comment by calculuschild** [comment]
Hm. Good point. Yeah, we don't want to have to open every file and parse out the text. That could become very costly.

*Maybe* updating based on the updatedAt timestamp... That's probably what I would do as well.

But I think you have a good argument that maybe this is rare enough a thing, and the stub does update itself if you open it in the edit page. Is the extra effort and potential cost here worth it just to make it match up on the user brew page? Maybe, but I probably wouldn't give this very high priority.

**Comment by 5e-Cleric** [comment]
First of all, there should be no duplication of data between stub and brew stored in GD, or as minimal as possible.

**Comment by G-Ambatte** [comment]
As I see it, the issue that there are two sources of the `tags` property, any of which might be the current "true" tags: **A.** MongoDB stub `tags` data; and **B.** code-fenced metadata in the `text` data (stored in **i.** MongoDB or **ii.** Google Drive).

The experimentation that I have done so far with tagging multiple brews at once from the UserPage has revolved around using Mongo's `updateMany` function to modify the `tags` field, causing **A** to be the most up to date data.  
But on opening the brew for editing, we can see that the tags are actually loaded from **B**, overwriting any prior value.

So I envision two ways to deal with this:
1. Remove `tags` from the code-fenced metadata and only load tags from the stub (effectively, **A** only).
2. Implement a system to identify and apply the "correct" set of tags from either **A** or **B**.

On 2, I would imagine some process like this:
![image](https://github.com/user-attachments/assets/ac9c8a1a-aee1-4376-ba04-14e382d746eb)

**Comment by G-Ambatte** [comment]
Addendum: this still does not resolve the situation where a user has updated their Google file outside of Homebrewery and manually changed the `tags` data, yet expects the User Page to detect and apply those tags without ever opening the brew in the editor.

**Comment by calculuschild** [comment]
> First of all, there should be no duplication of data between stub and brew stored in GD, or as minimal as possible.

I don't think we can eliminate all duplication. Google Drive needs to contain any data needed to rebuild the brew if HB MongoDB were to be lost, or if it were downloaded as a backup. So that includes Description, CSS, Renderer, Theme, etc.

Stubs needs anything that should be searchable. So IDs, Authors, Tags, Renderer (one example of duplication).

Does Tags need to be in both? Maybe it can just be in the Stub, if we agree it would be OK for a brew to lose all its tags when downloaded as a backup.

**Comment by G-Ambatte** [comment]
It would be MUCH easier to only store the tags in the stub.

The only use case I can come up with to store them in the code-fenced metadata is to update tags in an editor outside of Homebrewery by editing the Google Drive file directly, and I'm not even certain that's a use-case that we should choose to support.

**Comment by calculuschild** [comment]
> Implement a system to identify and apply the "correct" set of tags from either A or B.

This is essentially what we have already if we step through the logic:

If tags are edited via HB, they will be both update at the same time, thus be always equal, leading to the first leaf.

If it is edited externally in Google Drive, then by definition it has a GoogleID, so we never reach the second leaf. It will also always have a later updated time than in HB, so we never reach the third leaf, and always land on the fourth leaf.

The `metadata` field cannot be edited in any other way by a user, so we have only those two possibilities. Then we see that both leaves are equivalent to just applying the value from `metadata`.

**Comment by calculuschild** [comment]
> The only use case I can come up with...

I agree, even if people are editing their brew text externally, I don't actually imagine people are manually editing the metadata fence.

But instead I'm thinking of a case where someone is downloading .txt copies of all their brews to upload into a local deployment or something. Is it worth preserving the tags in `metadata` so the tags are still intact? Maybe not.

Like, if you use the "upload from text file" function, how much of the brew do we we actually need to preserve?

**Comment by G-Ambatte** [comment]
> we should first decide whether tags become stub-only

For the record, I am in favor of stub only tags.

**Comment by G-Ambatte** [comment]
As a thought:

1. Modify `splitTextStyleAndMetadata` to take a `loadTags` parameter which defaults to `false`. If set to `true`, then `tags` is included in the metadata extracted from the brew `text`.
2. Modify NewBrew's `handleFileChange` to call `splitTextStyleAndMetadata` with `loadTags=true`.

By leaving `mergeBrew` unchanged, `tags` will continue to be saved to the code-fenced metadata in the `text` field. When recovering from a downloaded file, `tags` will be restored. Total change is probably four to six lines of code.

**Comment by calculuschild** [comment]
> As a thought:
> 1. Modify splitTextStyleAndMetadata to take a loadTags parameter which defaults to false. If set to true, then tags is included in the metadata extracted from the brew text.
> 2. Modify NewBrew's handleFileChange to call splitTextStyleAndMetadata with loadTags=true.

We already load the metadata from the brew text metadata by default. So this would just be *removing* that behavior outside of `/new`. I don't think we want that?

**Comment by G-Ambatte** [comment]
Correct, this is for stub only tags. This would only load from text metadata when restoring a brew from a file.

**Comment by calculuschild** [comment]
But if we move to stub-only tags, the file being restored wouldn't have any tags to load anyway, right?

**Comment by G-Ambatte** [comment]
The tags would still be saved in the `text` metadata, as we have not changed `mergeBrew` to remove it - we just don't load them in `splitTextStyleAndMetadata` except in the case of restoring from a downloaded brew document.

**Comment by G-Ambatte** [comment]
Alternatively, we just remove `tags` from `splitTextStyleAndMetadata` and `mergeBrew` completely, and restored brews will have no tags. I don't believe that change would be significantly more complex.

**Comment by G-Ambatte** [comment]
I've put together a quick implementation in Draft PR #3932.

**Comment by G-Ambatte** [comment]
The quick implementation has revealed that we were not, in fact, recovering tags from a downloaded file during the NEW FROM FILE process.

**Comment by 5e-Cleric** [comment]
Always keep in mind older brews will have tags in the drive file, I say we keep tags to stubs, as remaking the tags is not that hard in case of upload. And we add a function to delete the tags from any brew that still has the tags in the codefence.

**Comment by G-Ambatte** [comment]
My understanding is that simply removing `tags` from `mergeBrew` will delete the tags from the code-fence when the brew is next updated.

**Comment by calculuschild** [comment]
> My understanding is that simply removing `tags` from `mergeBrew` will delete the tags from the code-fence when the brew is next updated.

I think I'm leaning toward just doing this, and then just stop loading them at all from splitTextStyleandMetadata. More straightforward if we just declare "tags aren't a vital part of a brew backup" since they mostly only make sense in the context of our online tool.

**Comment by G-Ambatte** [comment]
If the complaint is that the User Page does not extract the metadata values that are code-fenced into the `text` property, then IMO that is operating as intended and not a bug to be fixed.

> User Page info === MongoDB stub properties (`HomebrewModel.js:getByUser`) + Google Drive folder list properties (`GoogleActions.js:listGoogleBrews`)

For Google files, reading the code-fence from the `text` property requires reading the `file.data`, which I believe is not possible from a folder listing, and instead would require retrieving the full file data for every single document from Google Drive.
I suspect that such an approach would swiftly be rate-limited by Google.

**Comment by G-Ambatte** [comment]
Re-reading the original post, it seems to me that there is a misconception about how the brew metadata works. There are four sources of brew information; two from MongoDB and two from Google Drive:

A. MongoDB
1. MongoDB model properties
2. Code-fenced metadata inside the `textBin` property

B. Google Drive
1. Google Drive custom file properties (accessible via `drive.files.list`)
2. Code-fenced metadata inside the `file.data` property (accessible via `drive.files.get`)

In both cases, accessing the code-fenced metadata requires an additional step - the `textBin` field must be decompressed, or we need to use a different Google API call. As such, these actions are only performed when actually opening the brew.

Furthermore, in option A, both 1. and 2. are written in the same atomic update, so there is no possibility for the data to differ. While option B. is similar, the data can only differ when it is modified by some outside force, i.e. an edit from outside of Homebrewery.

---

To address the Proposals from the original post:

> 1. The userpage should display title and last-modified from the gdrive metadata, and whatever other valid metadata from gdrive file metadata (i.e. without needing to fetch the full file content)

As discussed, retrieving the code-fenced metadata is only possible by fetching the full file content. The only scenario in which it could possibly differ may be considered "unsupported usage" and not officially supported or encouraged.

With that said:

> 2. [optional] indicate if the gdrive last-modified is different from the mongo last-modified

We already fetch the `modifiedTime` property during `listGoogleBrews`, which is returned as the `brew.updatedAt` property. In `app.js`, we confirm the Google brew correlates to the stub by matching `editId` - it should be simple here to add a new property (`brew.metadataMismatch` or similar) when a disparity is detected (e.g. `brew.updatedAt != googleBrews[match].updatedAt`).

> 3. [optional-plus] one-click "refresh metadata" button if difference is detected (additionally this could even indicate gdrive brews that don't even have a mongo stub, with the one-click button thus creating the stub)

The new `brew.metadataMismatch` property could then be used to trigger the availability of this functionality, which could perform the retrieval of code-fenced metadata and update of the stub.

**Comment by G-Ambatte** [comment]
My current thinking is that the stub should be considered the "source of truth" for the metadata, and when a metadata property value is present, it should overwrite anything contained within the internal code-fence - effectively eliminating the possibility of editing metadata from outside Homebrewery.

But I'm open to being persuaded otherwise.

**Comment by dbolack-ab** [comment]
I agree with stubs being the Source of truth and the codefence being backups and/or user readable copies of that data.

**Comment by 5e-Cleric** [comment]
I agree too, stub > codefence, we should not specifically support outside editing.

Question, in case of a brew import, does the metadata get loaded (keep in mind, I made that function and don't know/remember the answer).

**Comment by G-Ambatte** [comment]
> I agree too, stub > codefence, we should not specifically support outside editing.
> 
> Question, in case of a brew import, does the metadata get loaded (keep in mind, I made that function and don't know/remember the answer).

Renderer, Theme, and Lang get loaded - and nothing else.

**Comment by dbolack-ab** [comment]
Based on above, should this issue be closed?

**Comment by 5e-Cleric** [comment]
I think it should be closed, i'm ready for a counter argument by @calculuschild, since he said:

> The issue here is when the Google Drive file is edited externally from The Homebrewery. In that case, the Google Drive file should take precedence if it doesn't match the stub in Mongo.

I say the stub should always take precedence, if no answer is given in.. 7 days, i say we close this.
